# Tarea 1: Explorando datos mundiales de COVID-19 

En esta actividad se pide que explore los datos del repositorio ["COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University"](https://github.com/CSSEGISandData/COVID-19)

Adjunto a este cuadernillo encontrará los datos al 26 de Abril de 2021 de dicho repositorio en el archivo `data.zip`. Luego de descomprimir debería observar cuatro archivos

- `covid19_confirmados.csv` : Número de casos diarios acumulado reportados como COVID-19 positivos (confirmados)
- `covid19_recuperados.csv` : Número de casos diarios acumulado reportados como recuperados 
- `covid19_decesos.csv` : Número de decesos diarios por COVID-19 acumulado 
- `population_by_country_2020.csv`:  Población por país más otros indicadores socioeconómicos al 2020 descargados desde: https://www.kaggle.com/tanuprabhu/population-by-country-2020

En los primeros tres archivos cada fila corresponde a un país o un estado/provincia de un país. Las columnas son las coordenadas geográficas (latitud y longitud) y el número de casos diarios para cada día desde Enero de 2020 a la fecha. Según quienes compilaron este repositorio, la información de casos recuperados debe analizarse con cautela pues podría estar considerablemente subestimada. 

A modo de ejemplo se muestra a continuación una gráfica de la series de tiempo del número de casos confirmados totales acumulados de Chile, Argentina y Bolivia 

<img src="ejemplo1.png">

A partir de esta información se puede calcular la serie de tiempo de los casos confirmados "nuevos" por día

<img src="ejemplo2.png">

En esta tarea se pide que procese, visualice, analice y discuta en base a estos datos. Escriba las rutinas de Python necesarias para resolver las actividades específicas que encontrará en este enunciado. **Use las librerías vistas en el curso**. Discuta con sus compañeros de grupo y complete con su análisis donde corresponda.

## Instrucciones generales 

1. Forme un grupo de **máximo tres estudiantes**
1. Versione su trabajo usando un **repositorio privado de github**. Agregue a sus compañeros y a su profesor (usuario github: phuijse) en la pestaña *Settings/Manage access*. No se aceptarán consultas de programación si no se cumple este requisito
1. Su tarea se evaluará en base al último commit antes de la fecha de entrega: **14:10 del 10 de Mayo de 2021**. La nota se calcula como ("pt totales" + 1)
1. [Sean leales y honestos](https://www.acm.org/about-acm/code-of-ethics-in-spanish), no copie ni comparta resultados con otros grupos

## Enunciado

### Preparación de datos (1.5pt)

1. Explore el archivo `covid19_confirmados.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    1. Utilice el nombre de país como índice principal
    1. Elimine las columnas `Lat` y `Long`
    1. Use TimeStamps para los "nombres de columna" asociados a fechas
    1. Reduzca la columna `Province/State`, el DataFrame resultante debe tener los totales a nivel país
1. Explore el archivo `mundiales.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    1. Utilice el nombre de país como índice principal. Algunos siglas o nombres podrían no calzar con el `DataFrame` anterior, modifíquelos manualmente o usando pandas. 
1. Haga un `merge` de los `DataFrame` anteriores. El objetivo es asignar un valor de `Population` a cada elemento del primer `DataFrame`. Indique las filas donde no se puede realizar el `merge` y luego descártelas de su `DataFrame` final
1. Repita el procedimiento anterior para los archivos `covid19_recuperados.csv` y `covid19_muertes.csv` 

### Análisis global (1.5pt)

1. Obtenga la cantidad total mundial a la fecha (26 de Abril) de los casos (a) confirmados, (b) cerrados (recuperados más decesos) y (c) activos (confirmados menos cerrados) por país. En base a estas cantidades calcule el porcentaje de 
    1. casos activos con respecto al total de confirmados
    1. decesos con respecto al total de casos cerrados
    1. recuperados con respecto al total de casos cerrados
1. Luego calcule las siguientes tasas en base a información a la fecha
    1. Tasa de incidencia: confirmados por 100.000 dividido población total
    1. Tasa de recuperación: recuperados por 100.000 dividido población total
    1. Tasa de mortalidad: decesos por 100.000 dividido población total
1. En base a lo anterior muestre los nombres y los valores de los 5 países con
    1. La mayor y menor tasa de incidencia, respectivamente
    1. La mayor y menor tasa de recuperación, respectivamente
    1. La mayor y menor tasa de mortalidad, respectivamente
1. Responda, analice y discuta: ¿En qué lugar/ranking se encuentra Chile con respecto a cada una de las métricas de los puntos anteriores? ¿A qué países de la [OCDE](https://es.wikipedia.org/wiki/Organizaci%C3%B3n_para_la_Cooperaci%C3%B3n_y_el_Desarrollo_Econ%C3%B3micos) se parece más? ¿A qué paises de latinoamérica se parece más? 

### Análisis temporal y comparativo (2.0pt)

1. Construya una interfaz gráfica usando `ipywidgets` y `matplotlib` que permita realizar comparaciones entre las series de tiempo de COVID-19 de dos países seleccionados arbitrariamente. A modo de simplificación considere sólo los países de la [OCDE](https://es.wikipedia.org/wiki/Organizaci%C3%B3n_para_la_Cooperaci%C3%B3n_y_el_Desarrollo_Econ%C3%B3micos). Requerimientos:
    1. La interfaz debe incluir controles para escoger dos países a partir de una lista
    1. La interfaz debe incluir tres [tabs](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#Tabs), el primero debe mostrar las series de tiempo de ambos países de los casos confirmados, el segundo la de recuperados y el tercero la de los decesos, respectivamente. Use una leyenda para indicar el nombre de los países.
    1. La interfaz debe tener un control que permita escoger entre casos acumulados y casos nuevos  
    1. La interfaz debe tener un control que permita escoger entre valores absolutos y valores relativos (tasas por 100.000 habitantes)
    1. La interfaz debe tener un control que permita escoger entre valores diarios y valores semanales (acumulados). Para esto último se recomienda usar `groupby` con una [frecuencia lunes a lunes](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#anchored-offsets)
    
En base a su interfaz:   

1. Describa en detalle la situación de Chile con respecto a la evolución temporal de las series de tiempo. En particular destaque patrones y comportamientos que considere relevantes y compare lo que observa en las distintas series de tiempo. Considerando la evolución a nivel de semana, conteste ¿En qué fechas ocurre un aumento o disminución notoria de los casos? ¿En qué fechas hay máximos y mínimos?
1. Considerando métricas que sean relativas compare el caso de Chile contra cinco países de la OCDE seleccionados por ustedes. Destaque semejanzas y diferencias considerando las cantidades y los posibles desfases temporales existentes.


### Dashboard interactivo (1.0pt)

Implemente un *dashboard web* interactivo a partir de la interfaz desarrollada en el punto anterior utilizando la librería [voila](https://github.com/voila-dashboards/voila). Sirva su interfaz en la nube, para esto puede considerar los servicios gratuitos [binder](https://mybinder.org/) o [heroku](https://www.heroku.com/free). Revise la documentación de voila [con respecto a *deployment* en la nube](https://voila.readthedocs.io/en/stable/deploy.html) y este repositorio con [un ejemplo en base a heroku](https://github.com/voila-dashboards/voila-heroku). Entregue el link de su dashboard web

# Desarrollo

## Preparación de Datos

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import matplotlib.dates as mdates
from ipywidgets import interact, interactive, fixed, interact_manual
import ipywidgets as widgets

### covid19_confirmados.csv

Creamos el DataFrame, hacemos index la columna 'Country/Region' y usamos TimeStamp para las columnas que son fechas

In [2]:
covid19_confirmados = pd.read_csv("data/covid19_confirmados.csv",index_col='Country/Region', parse_dates=True, infer_datetime_format=True)

Eliminando Lat y Long

In [3]:
covid19_confirmados.drop(labels = ["Lat","Long"], axis=1, inplace =True)

Agrupando por ciudad

In [4]:
covid19_confirmados = covid19_confirmados.groupby(by=covid19_confirmados.index)[covid19_confirmados.columns].sum()
covid19_confirmados[3:4:15]

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Andorra               0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  ...  4/16/21  4/17/21  4/18/21  \
Country/Region                             ...                              
Andorra               0        0        0  ...    12712    12771    12805   

                4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  4/24/21  4/25/21  
Country/Region                                                                 
Andorra           12805    12874    12917    12942    13007    13024    13060  

[1 rows x 460 columns]

## population_by_country_2020.csv

In [5]:
 mundiales = pd.read_csv("data/population_by_country_2020.csv", index_col = 'Country (or dependency)')

In [6]:
mundiales.drop(mundiales.columns[[1,2,3,4,5,6,7,8,9]], axis='columns', inplace=True)
mundiales.head()

Population (2020)
Country (or dependency)                   
China                           1440297825
India                           1382345085
US                               331341050
Indonesia                        274021604
Pakistan                         221612785

In [7]:
confirmados_mundial = pd.merge(covid19_confirmados, mundiales, how = 'inner', left_index = True, right_index = True)
confirmados_mundial.head(10)

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Angola                     0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   
Argentina                  0        0        0        0        0        0   
Armenia                    0        0        0        0        0        0   
Australia                  0        0        0        0        4        5   
Austria                    0        0        0        0        0        0   

                     1/28/20  1/29/20  1/30/20  1/31/20  ...  4/17/21  \
Afghanistan                0        0        0        0  ...    57721   
Albania                    0        0        0        0  ...   129456   
Algeria                    0        0        0        0  ...   119486   
Andorra                    0        0        0        0  ...    12771   
Angola                     0        0        0        0  ...    24300   
Antigua and Barbuda        0        0        0        0  ...     1216   
Argentina                  0        0        0        0  ...  2677747   
Armenia                    0        0        0        0  ...   207973   
Australia                  5        6        9        9  ...    29519   
Austria                    0        0        0        0  ...   591347   

                     4/18/21  4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  \
Afghanistan            57793    57898    58037    58214    58312    58542   
Albania               129594   129694   129842   129980   130114   130270   
Algeria               119642   119805   119992   120174   120363   120562   
Andorra                12805    12805    12874    12917    12942    13007   
Angola                 24389    24518    24661    24883    25051    25279   
Antigua and Barbuda     1216     1217     1217     1217     1217     1222   
Argentina            2694014  2714475  2743620  2769552  2796768  2824652   
Armenia               208520   208818   209485   210518   211399   212114   
Australia              29533    29556    29576    29594    29638    29653   
Austria               593423   595540   597566   600089   602494   604823   

                     4/24/21  4/25/21  Population (2020)  
Afghanistan            58730    58843           39074280  
Albania               130409   130537            2877239  
Algeria               120736   120922           43984569  
Andorra                13024    13060              77287  
Angola                 25492    25609           33032075  
Antigua and Barbuda     1227     1227              98069  
Argentina            2845872  2860884           45267449  
Armenia               212878   213288            2964219  
Australia              29663    29680           25550683  
Austria               606954   608979            9015361  

[10 rows x 461 columns]

In [8]:
nan_values = confirmados_mundial[confirmados_mundial.isna().any(axis=1)]
nan_values

Empty DataFrame
Columns: [1/22/20, 1/23/20, 1/24/20, 1/25/20, 1/26/20, 1/27/20, 1/28/20, 1/29/20, 1/30/20, 1/31/20, 2/1/20, 2/2/20, 2/3/20, 2/4/20, 2/5/20, 2/6/20, 2/7/20, 2/8/20, 2/9/20, 2/10/20, 2/11/20, 2/12/20, 2/13/20, 2/14/20, 2/15/20, 2/16/20, 2/17/20, 2/18/20, 2/19/20, 2/20/20, 2/21/20, 2/22/20, 2/23/20, 2/24/20, 2/25/20, 2/26/20, 2/27/20, 2/28/20, 2/29/20, 3/1/20, 3/2/20, 3/3/20, 3/4/20, 3/5/20, 3/6/20, 3/7/20, 3/8/20, 3/9/20, 3/10/20, 3/11/20, 3/12/20, 3/13/20, 3/14/20, 3/15/20, 3/16/20, 3/17/20, 3/18/20, 3/19/20, 3/20/20, 3/21/20, 3/22/20, 3/23/20, 3/24/20, 3/25/20, 3/26/20, 3/27/20, 3/28/20, 3/29/20, 3/30/20, 3/31/20, 4/1/20, 4/2/20, 4/3/20, 4/4/20, 4/5/20, 4/6/20, 4/7/20, 4/8/20, 4/9/20, 4/10/20, 4/11/20, 4/12/20, 4/13/20, 4/14/20, 4/15/20, 4/16/20, 4/17/20, 4/18/20, 4/19/20, 4/20/20, 4/21/20, 4/22/20, 4/23/20, 4/24/20, 4/25/20, 4/26/20, 4/27/20, 4/28/20, 4/29/20, 4/30/20, ...]
Index: []

[0 rows x 461 columns]

Al realizar un inner join no existen filas con valores NaN

### covid19_recuperados.csv

In [9]:
covid19_recuperados = pd.read_csv("data/covid19_recuperados.csv",index_col='Country/Region', parse_dates=True, infer_datetime_format=True)

In [10]:
covid19_recuperados.drop(labels = ["Lat","Long"], axis=1, inplace =True)

In [11]:
covid19_recuperados = covid19_recuperados.groupby(by=covid19_recuperados.index)[covid19_recuperados.columns].sum()
covid19_recuperados.head()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Afghanistan           0        0        0        0        0        0        0   
Albania               0        0        0        0        0        0        0   
Algeria               0        0        0        0        0        0        0   
Andorra               0        0        0        0        0        0        0   
Angola                0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  ...  4/16/21  4/17/21  4/18/21  \
Country/Region                             ...                              
Afghanistan           0        0        0  ...    52105    52116    52168   
Albania               0        0        0  ...   100600   101142   101584   
Algeria               0        0        0  ...    83169    83286    83397   
Andorra               0        0        0  ...    12105    12159    12203   
Angola                0        0        0  ...    22203    22576    22597   

                4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  4/24/21  4/25/21  
Country/Region                                                                 
Afghanistan       52244    52272    52301    52348    52363    52392    52489  
Albania          102171   102601   103066   103582   104278   105016   105728  
Algeria           83514    83636    83765    83900    84038    84167    84299  
Andorra           12203    12285    12334    12375    12423    12458    12491  
Angola            22600    22647    22882    22901    23089    23092    23092  

[5 rows x 460 columns]

In [12]:
recuperados_mundial = pd.merge(covid19_recuperados, mundiales, how = 'inner', left_index = True, right_index = True)
recuperados_mundial.head(10)

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Angola                     0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   
Argentina                  0        0        0        0        0        0   
Armenia                    0        0        0        0        0        0   
Australia                  0        0        0        0        0        0   
Austria                    0        0        0        0        0        0   

                     1/28/20  1/29/20  1/30/20  1/31/20  ...  4/17/21  \
Afghanistan                0        0        0        0  ...    52116   
Albania                    0        0        0        0  ...   101142   
Algeria                    0        0        0        0  ...    83286   
Andorra                    0        0        0        0  ...    12159   
Angola                     0        0        0        0  ...    22576   
Antigua and Barbuda        0        0        0        0  ...      986   
Argentina                  0        0        0        0  ...  2343808   
Armenia                    0        0        0        0  ...   187256   
Australia                  0        0        2        2  ...    23158   
Austria                    0        0        0        0  ...   552619   

                     4/18/21  4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  \
Afghanistan            52168    52244    52272    52301    52348    52363   
Albania               101584   102171   102601   103066   103582   104278   
Algeria                83397    83514    83636    83765    83900    84038   
Andorra                12203    12203    12285    12334    12375    12423   
Angola                 22597    22600    22647    22882    22901    23089   
Antigua and Barbuda      986      995      995      995      995      995   
Argentina            2364171  2386383  2407853  2424675  2449793  2474514   
Armenia               187466   189017   189688   190271   191158   192281   
Australia              23160    23173    23177    23181    23183    23191   
Austria               555218   557467   560492   563256   565513   568213   

                     4/24/21  4/25/21  Population (2020)  
Afghanistan            52392    52489           39074280  
Albania               105016   105728            2877239  
Algeria                84167    84299           43984569  
Andorra                12458    12491              77287  
Angola                 23092    23092           33032075  
Antigua and Barbuda     1002     1002              98069  
Argentina            2496277  2518167           45267449  
Armenia               193360   194080            2964219  
Australia              23195    23209           25550683  
Austria               570684   573155            9015361  

[10 rows x 461 columns]

In [13]:
nan_values = recuperados_mundial[recuperados_mundial.isna().any(axis=1)]
nan_values

Empty DataFrame
Columns: [1/22/20, 1/23/20, 1/24/20, 1/25/20, 1/26/20, 1/27/20, 1/28/20, 1/29/20, 1/30/20, 1/31/20, 2/1/20, 2/2/20, 2/3/20, 2/4/20, 2/5/20, 2/6/20, 2/7/20, 2/8/20, 2/9/20, 2/10/20, 2/11/20, 2/12/20, 2/13/20, 2/14/20, 2/15/20, 2/16/20, 2/17/20, 2/18/20, 2/19/20, 2/20/20, 2/21/20, 2/22/20, 2/23/20, 2/24/20, 2/25/20, 2/26/20, 2/27/20, 2/28/20, 2/29/20, 3/1/20, 3/2/20, 3/3/20, 3/4/20, 3/5/20, 3/6/20, 3/7/20, 3/8/20, 3/9/20, 3/10/20, 3/11/20, 3/12/20, 3/13/20, 3/14/20, 3/15/20, 3/16/20, 3/17/20, 3/18/20, 3/19/20, 3/20/20, 3/21/20, 3/22/20, 3/23/20, 3/24/20, 3/25/20, 3/26/20, 3/27/20, 3/28/20, 3/29/20, 3/30/20, 3/31/20, 4/1/20, 4/2/20, 4/3/20, 4/4/20, 4/5/20, 4/6/20, 4/7/20, 4/8/20, 4/9/20, 4/10/20, 4/11/20, 4/12/20, 4/13/20, 4/14/20, 4/15/20, 4/16/20, 4/17/20, 4/18/20, 4/19/20, 4/20/20, 4/21/20, 4/22/20, 4/23/20, 4/24/20, 4/25/20, 4/26/20, 4/27/20, 4/28/20, 4/29/20, 4/30/20, ...]
Index: []

[0 rows x 461 columns]

### covid19_decesos.csv

In [14]:
covid19_decesos = pd.read_csv("data/covid19_decesos.csv",index_col='Country/Region', parse_dates=True, infer_datetime_format=True)

In [15]:
covid19_decesos.drop(labels = ["Lat","Long"], axis=1, inplace =True)

In [16]:
covid19_decesos = covid19_decesos.groupby(by=covid19_decesos.index)[covid19_decesos.columns].sum()
covid19_decesos.head()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Afghanistan           0        0        0        0        0        0        0   
Albania               0        0        0        0        0        0        0   
Algeria               0        0        0        0        0        0        0   
Andorra               0        0        0        0        0        0        0   
Angola                0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  ...  4/16/21  4/17/21  4/18/21  \
Country/Region                             ...                              
Afghanistan           0        0        0  ...     2535     2539     2539   
Albania               0        0        0  ...     2337     2340     2342   
Algeria               0        0        0  ...     3148     3152     3155   
Andorra               0        0        0  ...      123      123      123   
Angola                0        0        0  ...      560      561      561   

                4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  4/24/21  4/25/21  
Country/Region                                                                 
Afghanistan        2546     2549     2557     2561     2565     2572     2582  
Albania            2347     2353     2358     2364     2367     2372     2378  
Algeria            3160     3165     3172     3181     3190     3198     3207  
Andorra             123      123      123      123      123      124      124  
Angola              563      565      570      572      574      577      579  

[5 rows x 460 columns]

In [17]:
decesos_mundial = pd.merge(covid19_decesos, mundiales, how = 'inner', left_index = True, right_index = True)
decesos_mundial.head(10)

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Angola                     0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   
Argentina                  0        0        0        0        0        0   
Armenia                    0        0        0        0        0        0   
Australia                  0        0        0        0        0        0   
Austria                    0        0        0        0        0        0   

                     1/28/20  1/29/20  1/30/20  1/31/20  ...  4/17/21  \
Afghanistan                0        0        0        0  ...     2539   
Albania                    0        0        0        0  ...     2340   
Algeria                    0        0        0        0  ...     3152   
Andorra                    0        0        0        0  ...      123   
Angola                     0        0        0        0  ...      561   
Antigua and Barbuda        0        0        0        0  ...       31   
Argentina                  0        0        0        0  ...    59164   
Armenia                    0        0        0        0  ...     3859   
Australia                  0        0        0        0  ...      910   
Austria                    0        0        0        0  ...     9870   

                     4/18/21  4/19/21  4/20/21  4/21/21  4/22/21  4/23/21  \
Afghanistan             2539     2546     2549     2557     2561     2565   
Albania                 2342     2347     2353     2358     2364     2367   
Algeria                 3155     3160     3165     3172     3181     3190   
Andorra                  123      123      123      123      123      123   
Angola                   561      563      565      570      572      574   
Antigua and Barbuda       31       31       31       31       31       31   
Argentina              59228    59476    59792    60083    60620    61176   
Armenia                 3878     3895     3919     3944     3969     3984   
Australia                910      910      910      910      910      910   
Austria                 9898     9922     9959     9997    10026    10055   

                     4/24/21  4/25/21  Population (2020)  
Afghanistan             2572     2582           39074280  
Albania                 2372     2378            2877239  
Algeria                 3198     3207           43984569  
Andorra                  124      124              77287  
Angola                   577      579           33032075  
Antigua and Barbuda       31       31              98069  
Argentina              61474    61644           45267449  
Armenia                 4001     4018            2964219  
Australia                910      910           25550683  
Austria                10070    10087            9015361  

[10 rows x 461 columns]

In [18]:
nan_values = decesos_mundial[decesos_mundial.isna().any(axis=1)]
nan_values

Empty DataFrame
Columns: [1/22/20, 1/23/20, 1/24/20, 1/25/20, 1/26/20, 1/27/20, 1/28/20, 1/29/20, 1/30/20, 1/31/20, 2/1/20, 2/2/20, 2/3/20, 2/4/20, 2/5/20, 2/6/20, 2/7/20, 2/8/20, 2/9/20, 2/10/20, 2/11/20, 2/12/20, 2/13/20, 2/14/20, 2/15/20, 2/16/20, 2/17/20, 2/18/20, 2/19/20, 2/20/20, 2/21/20, 2/22/20, 2/23/20, 2/24/20, 2/25/20, 2/26/20, 2/27/20, 2/28/20, 2/29/20, 3/1/20, 3/2/20, 3/3/20, 3/4/20, 3/5/20, 3/6/20, 3/7/20, 3/8/20, 3/9/20, 3/10/20, 3/11/20, 3/12/20, 3/13/20, 3/14/20, 3/15/20, 3/16/20, 3/17/20, 3/18/20, 3/19/20, 3/20/20, 3/21/20, 3/22/20, 3/23/20, 3/24/20, 3/25/20, 3/26/20, 3/27/20, 3/28/20, 3/29/20, 3/30/20, 3/31/20, 4/1/20, 4/2/20, 4/3/20, 4/4/20, 4/5/20, 4/6/20, 4/7/20, 4/8/20, 4/9/20, 4/10/20, 4/11/20, 4/12/20, 4/13/20, 4/14/20, 4/15/20, 4/16/20, 4/17/20, 4/18/20, 4/19/20, 4/20/20, 4/21/20, 4/22/20, 4/23/20, 4/24/20, 4/25/20, 4/26/20, 4/27/20, 4/28/20, 4/29/20, 4/30/20, ...]
Index: []

[0 rows x 461 columns]

## Análisis Global

### Total de Casos a la Fecha a Nivel Mundial

#### Casos Confirmados

In [19]:
total_confirmados = confirmados_mundial['4/25/21'].sum()
total_confirmados

146639797

#### Casos Recuperados

In [20]:
total_recuperados = recuperados_mundial['4/25/21'].sum()
total_recuperados

84382402

#### Decesos

In [21]:
total_decesos = decesos_mundial['4/25/21'].sum()
total_decesos

3100569

### Porcentajes a Nivel Mundial

#### Casos activos respecto al total de confirmados

In [22]:
activos = (total_confirmados - total_recuperados) - total_decesos
poblacion = mundiales['Population (2020)'].sum()

porcentaje_activos = (activos*100)/total_confirmados
porcentaje_activos

40.34159021646764

El porcentaje de casos activos respecto al total de casos confirmados es de **40.34%**

#### Decesos respecto al total de casos cerrados

In [23]:
cerrados = total_recuperados + total_decesos

porcentaje_decesos = (total_decesos/cerrados)*100
porcentaje_decesos

3.544197190102288

El porcentaje de decesos a nivel mundial respecto al total de casos cerrados es de un **3.54%**

#### Recuperados respecto al total de casos cerrados

In [24]:
porcentaje_recuperados = (total_recuperados/cerrados)*100
porcentaje_recuperados

96.4558028098977

El porcentaje de casos recuperados respecto a los casos cerrados a nivel mundial es de **96.46%**

### Tasas de Incidencia, Recuperación y Mortalidad

#### Incidencia

In [25]:
t_incidencia = (total_confirmados * 100000)/poblacion
t_incidencia

1877.9630804461922

La tasa de incidencia es de **1878 personas contagiadas por cada 100000 personas.**

#### Recuperación

In [26]:
t_recuperacion = (total_recuperados * 100000)/poblacion
t_recuperacion

1080.655039336756

**Por cada 100000 personas existen 1081** que se han recuperado del Covid19.

#### Mortalidad

In [27]:
t_mortalidad = (total_decesos * 100000)/poblacion
t_mortalidad

39.70787077928081

**40 de cada 100000 personas** mueren debido al Covid19 a nivel mundial.

### Top 5

In [28]:
df_incidencia = confirmados_mundial['4/25/21'].mul(100000).div(confirmados_mundial['Population (2020)']).sort_values(0)
df_recuperacion = recuperados_mundial['4/25/21'].mul(100000).div(recuperados_mundial['Population (2020)']).sort_values(0)
df_mortalidad = decesos_mundial['4/25/21'].mul(100000).div(decesos_mundial['Population (2020)']).sort_values(0)

**Mayor tasa de incidencia**

In [29]:
df_incidencia.tail()

Slovenia      11401.118524
San Marino    14868.607118
Czechia       15116.787605
Montenegro    15389.759266
Andorra       16898.055300
dtype: float64

**Menor tasa de incidencia**

In [30]:
df_incidencia.head()

Tanzania           0.848158
Micronesia         0.867822
Vanuatu            1.297282
Samoa              1.510247
Solomon Islands    2.899933
dtype: float64

**Mayor tasa de recuperación**

In [31]:
df_recuperacion.tail()

Slovenia      10695.487085
Czechia       14227.759190
San Marino    14255.833137
Montenegro    14744.777735
Andorra       16161.838343
dtype: float64

**Menor tasa de recuperación**

In [32]:
df_recuperacion.head()

Serbia      0.000000
Sweden      0.000000
Belgium     0.000000
US          0.000000
Tanzania    0.304937
dtype: float64

**Mayor tasa de mortalidad**

In [33]:
df_mortalidad.tail()

Montenegro                233.250541
Bosnia and Herzegovina    250.308805
San Marino                262.196559
Czechia                   270.357539
Hungary                   275.735780
dtype: float64

**Menor tasa de mortalidad**

In [34]:
df_mortalidad.head()

Saint Kitts and Nevis    0.0
Marshall Islands         0.0
Laos                     0.0
Samoa                    0.0
Holy See                 0.0
dtype: float64

### Discusión

¿En qué lugar/ranking se encuentra Chile con respecto a cada una de las métricas de los puntos anteriores?

Incidencia:
Recuperación:
Mortalidad:

In [35]:
rank = (df_incidencia.index == "Chile")
print(np.where(rank==True)[0][0]+1)

144


In [36]:
rank = (df_recuperacion.index == "Chile")
print(np.where(rank==True)[0][0]+1)

158


In [37]:
rank = (df_mortalidad.index == "Chile")
print(np.where(rank==True)[0][0]+1)

154


### Comparación respecto a países de la OCDE

In [38]:
paises_ocde=['Austria','Australia','Belgium','Canada','Chile','Colombia','Czechia','Denmark','Estonia',
        'Finland','France','Germany','Greece','Hungary','Iceland','Ireland','Israel','Italy','Japan','South Korea',
        'Latvia','Lithuania','Luxembourg','Mexico','Netherlands','New Zealand','Norway','Poland','Portugal',
        'Slovakia','Slovenia','Spain','Sweden','Switzerland','Turkey','United Kingdom','US']

 ¿A qué países de la OCDE se parece más?

In [39]:
df_incidencia_ocde = df_incidencia[df_incidencia.index == "Chile"]
for i in paises_ocde:
    df_incidencia_ocde = df_incidencia_ocde.append(df_incidencia[df_incidencia.index == i]).sort_values(0)
df_incidencia_ocde

New Zealand          53.861849
Australia           116.161278
South Korea         232.827688
Japan               449.831973
Finland            1548.183522
Mexico             1802.634203
Iceland            1870.455583
Norway             2037.581002
Canada             3140.237094
Greece             3198.876966
Germany            3944.475319
Denmark            4284.427635
Ireland            4985.237304
Colombia           5442.660276
Turkey             5479.561731
Chile              6108.958634
Chile              6108.958634
Latvia             6125.983315
United Kingdom     6505.599362
Italy              6555.721976
Austria            6754.904213
Slovakia           6957.938019
Poland             7290.989212
Spain              7418.235347
Switzerland        7460.624549
Hungary            7969.338803
Portugal           8187.700052
Belgium            8401.277033
France             8513.341643
Netherlands        8670.193303
Lithuania          8886.437794
Estonia            9061.252302
Sweden  

En términos de **incidencia** se parece más a **Turquía y Latvia.**

In [40]:
df_recuperacion_ocde = df_recuperacion[df_recuperacion.index == "Chile"]
for i in paises_ocde:
    df_recuperacion_ocde = df_recuperacion_ocde.append(df_recuperacion[df_recuperacion.index == i]).sort_values(0)
df_recuperacion_ocde

Belgium               0.000000
US                    0.000000
Sweden                0.000000
United Kingdom       21.269118
New Zealand          52.640069
Australia            90.835145
Netherlands         136.072923
South Korea         212.278505
Spain               321.604997
Norway              331.540727
Japan               396.695852
Ireland             472.260745
France              517.912009
Finland             829.989768
Greece              900.370427
Mexico             1432.791601
Iceland            1822.742867
Canada             2859.939046
Germany            3480.491673
Switzerland        3665.060126
Denmark            4077.345931
Slovakia           4675.730832
Turkey             4821.151884
Hungary            4993.867533
Colombia           5077.274420
Latvia             5591.455200
Italy              5595.443936
Chile              5745.315717
Chile              5745.315717
Austria            6357.537984
Poland             6446.775974
Portugal           7777.972604
Lithuani

En términos de **recuperación** se parece más a **Italia y Austria.**

In [41]:
df_mortalidad_ocde = df_mortalidad[df_mortalidad.index == "Chile"]
for i in paises_ocde:
    df_mortalidad_ocde = df_mortalidad_ocde.append(df_mortalidad[df_mortalidad.index == i]).sort_values(0)
df_mortalidad_ocde

New Zealand         0.538411
South Korea         3.543501
Australia           3.561549
Japan               7.862671
Iceland             8.488766
Norway             13.557838
Finland            16.293060
Denmark            42.720738
Turkey             45.396638
Canada             63.360782
Israel             73.192229
Estonia            86.154069
Greece             96.092390
Germany            97.423420
Ireland            98.498828
Netherlands       100.813556
Latvia            111.665484
Austria           111.886812
Switzerland       121.976340
Luxembourg        125.855499
Chile             135.056325
Chile             135.056325
Sweden            137.711960
Colombia          139.969109
Lithuania         141.713377
France            157.762156
Spain             165.941728
Mexico            166.411403
Portugal          166.463734
US                172.692155
Poland            172.876025
United Kingdom    187.909092
Italy             197.263559
Slovenia          202.406073
Belgium       

En términos de **mortalidad** se parece más a **Luxemburgo y Suecia.**

### Comparación respecto a latinoamérica

In [42]:
paises_latam=['Argentina', 'Colombia',
              'Bolivia', 'Ecuador', 'Peru', 'Venezuela',
              'Brazil', 'Guatemala', 'Uruguay', 'Nicaragua', 'Paraguay',
             'Panama', 'Puerto Rico', 'Cuba']

In [43]:
df_incidencia_latam = df_incidencia[df_incidencia.index == "Chile"]
for i in paises_latam:
    df_incidencia_latam = df_incidencia_latam.append(df_incidencia[df_incidencia.index == i]).sort_values(0)
df_incidencia_latam

Nicaragua     102.966598
Venezuela     673.847102
Cuba          896.242788
Guatemala    1233.616869
Ecuador      2110.749415
Bolivia      2544.049007
Paraguay     3736.691424
Uruguay      5318.567415
Peru         5329.996229
Colombia     5442.660276
Chile        6108.958634
Argentina    6319.958520
Brazil       6738.395440
Panama       8389.786552
dtype: float64

En términos de **incidencia**, se parece mas a **Colombia y Argentina**

In [44]:
df_recuperacion_latam = df_recuperacion[df_recuperacion.index == "Chile"]
for i in paises_latam:
    df_recuperacion_latam = df_recuperacion_latam.append(df_recuperacion[df_recuperacion.index == i]).sort_values(0)
df_recuperacion_latam

Nicaragua      63.647970
Venezuela     609.033678
Cuba          839.661960
Guatemala    1094.278670
Ecuador      1801.148864
Bolivia      2099.390207
Paraguay     3073.037724
Uruguay      4444.649670
Colombia     5077.274420
Peru         5107.655742
Argentina    5562.864830
Chile        5745.315717
Brazil       5927.281874
Panama       8155.544604
dtype: float64

En términos de **recuperación**, se parece mas a **Brasil y Argentina**

In [45]:
df_mortalidad_latam = df_mortalidad[df_mortalidad.index == "Chile"]
for i in paises_latam:
    df_mortalidad_latam = df_mortalidad_latam.append(df_mortalidad[df_mortalidad.index == i]).sort_values(0)
df_mortalidad_latam

Nicaragua      2.726694
Cuba           5.218363
Venezuela      7.202274
Guatemala     41.232221
Uruguay       66.919037
Paraguay      82.545733
Ecuador      102.761106
Bolivia      109.604898
Chile        135.056325
Argentina    136.177322
Colombia     139.969109
Panama       143.517688
Peru         180.706864
Brazil       183.626235
dtype: float64

En términos de **mortalidad**, se parece mas a **Bolivia y Argentina**

## Análisis Temporal y Comparativo

In [46]:
def f(x,y):
    titleNames=['Confirmados','Recuperados','Decesos']
    out1 = widgets.Output()
    out2 = widgets.Output()
    out3 = widgets.Output()
    data1 = pd.DataFrame(np.random.normal(size = 50))
    data2 = pd.DataFrame(np.random.normal(size = 100))
    tab = widgets.Tab(children = [out1,out2,out3])
    for i in range(len(titleNames)):
        tab.set_title(i, titleNames[i])
    display(tab)
    with out1:
        df1 = covid19_confirmados.groupby('Country/Region').sum().reset_index()
        
        k = df1[df1['Country/Region']==x].loc[:,'1/24/20':]
        country1_confirmados = k.values.tolist()[0]
        k = df1[df1['Country/Region']==y].loc[:,'1/24/20':]
        country2_confirmados = k.values.tolist()[0]
        
        dates = list(covid19_confirmados.columns[:])
        dates = list(pd.to_datetime(dates))
        dates_country = dates[2:]
        fig1, axes1_1= plt.subplots(figsize= (6,6))
        plt.xticks(rotation = 60)
        plt.xlabel("Fecha")
        plt.ylabel("Casos confirmados")
        plt.title("Total Confirmados") 
        axes1_1 = plt.plot_date(y=country1_confirmados,x=dates_country,label = x,color = 'r')
        axes1_1 = plt.plot_date(y=country2_confirmados,x=dates_country,label = y,color = 'b')
        plt.legend()
        plt.show(fig1)
    with out2:
        df1 = covid19_recuperados.groupby('Country/Region').sum().reset_index()
        
        k = df1[df1['Country/Region']==x].loc[:,'1/24/20':]
        country1_recuperados = k.values.tolist()[0]
        k = df1[df1['Country/Region']==y].loc[:,'1/24/20':]
        country2_recuperados = k.values.tolist()[0]
        
        dates = list(covid19_recuperados.columns[:])
        dates = list(pd.to_datetime(dates))
        dates_country = dates[2:]
        fig2, axes2_1= plt.subplots(figsize= (6,6))
        plt.xticks(rotation = 60)
        plt.xlabel("Fecha")
        plt.ylabel("Casos Recuperados")
        plt.title("Total Recuperados") 
        axes2_1 = plt.plot_date(y=country1_recuperados,x=dates_country,label = x,color = 'r',linestyle ='-')
        axes2_1 = plt.plot_date(y=country2_recuperados,x=dates_country,label = y,color = 'b',linestyle ='-')
        plt.legend()
        plt.show(fig2)
    with out3:
        df1 = covid19_decesos.groupby('Country/Region').sum().reset_index()
        
        k = df1[df1['Country/Region']==x].loc[:,'1/24/20':]
        country1_decesos = k.values.tolist()[0]
        k = df1[df1['Country/Region']==y].loc[:,'1/24/20':]
        country2_decesos = k.values.tolist()[0]
        
        dates = list(covid19_decesos.columns[:])
        dates = list(pd.to_datetime(dates))
        dates_country = dates[2:]
        fig3, axes3_1= plt.subplots(figsize= (6,6))
        plt.xticks(rotation = 60)
        plt.xlabel("Fecha")
        plt.ylabel("Casos Decesos")
        plt.title("Total Decesos") 
        axes3_1 = plt.plot_date(y=country1_decesos,x=dates_country,label = x,color = 'r',linestyle ='-')
        axes3_1 = plt.plot_date(y=country2_decesos,x=dates_country,label = y,color = 'b',linestyle ='-')
        plt.legend()
        plt.show(fig3)
    return

Interactiveplot = interact(f, x=paises_ocde, y=paises_ocde)

interactive(children=(Dropdown(description='x', options=('Austria', 'Australia', 'Belgium', 'Canada', 'Chile',…

Se analizaron las tres series de tiempo correspondientes a **Confirmados, Recuperados y Decesos** concluyendo lo siguiente:
En el mes de Junio de 2020 encontramos la primera alza en confirmados, donde aproximadamente dos semanas después también comenzaron a subir en magnitud similar, los casos de recuperados y decesos. Posterior a este peak, se mantuvo un alza relativamente constante en confirmados y recuperados, mientras decesos subió a razón un poco menor que en el peak.
Luego, en Marzo de 2021 encontramos el segundo peak, donde confirmados y recuperados tuvo un mayor alza respecto de Decesos, se considera que la cantidad de Decesos no se elevó en la misma razón que Marzo de 2020 gracias a la vacunación en el país.

Comparación del país Chile respecto a otros países de la OCDE:

Chile - New Zealand: La comparación es drástica, el país oceánico practicamente no posee casos, el gráfico de alzas prácticamente se mantiene paralelo al eje X, con Y=0 a lo largo de todo el tiempo. El peak de Chile supera más de cien veces el peak de New Zealand.

Chile - United States: Estados Unidos presenta considerablemente mayor cantidad de casos que en Chile, considerando su población, es considerado normal, de hecho, el comportamiento es similar al país sudamericano, ya que para sus casos confirmados posee dos peak importantes en el mes de abril 2020 y en noviembre de 2020.

Chile - Colombia: Para estos países, en casos Confirmados existe similitud hasta mayor de 2020 y después de Enero 2021, en el periodo intermedio a estas fechas, Colombia tuvo un alza mucho mayor que se ha mantenido constante, mientras Chile ha rebajado sus niveles de confirmados.

Chile - Italia: Los casos confirmados de ambos países se mantuvieron similares hasta octubre de 2020, con la diferencia que Italia tuvo su peak inicial a comienzos de Abril. Luego de esta fecha, Italia tuvo un peak muy grande y constante hasta la fecha actual. Sin embargo, respecto a la tasa de recuperados Chile tuvo mucho mejores números, ya que cuando incuso Italia tenia mas caso confirmados, Chile poseía un mayor número de Recuperados. Además, en casos de Decesos, Italia supera con creces al país latinoamericano.

Chile - Francia: Ambos países venían en una situación muy similar respecto a confirmados hasta finales de Septiembre de 2020, sin embargo, desde esta fecha, Francia tuvo un incremento muy grande que se mantuvo constante hasta la fecha. Respecto a recuperados y Decesos, el caso es similar a Italia, donde Chile supera en cantidad de recuperados y posee menor cantidad de Decesos.

## Dashboard Interactivo en Voila

**Link:**